### ISSUE TITLE
Add server sided bots

### ISSUE THREAD

**Issue by oy** [teeworlds/teeworlds#1966]
Add server sided bots with which players can learn to play or have something to play with on an empty server.

**Comment by AssassinTee** [comment]
Can you leave some instructions here? What should the bots be able to do, what do you mean by 'which players can learn to play', what kind of bots do you want? You could try machine learning (Random Forests, Neural-Network) or simple behavior trees.

Should they be based on the already implemented dummies?

**Comment by oy** [comment]
The bots should be able to play the game and should have some adjustable difficulty.

**Comment by MegaManSec** [comment]
If the guy is still active, https://www.teeworlds.com/forum/viewtopic.php?id=11397 had some nice stuff

**Comment by Dune-jr** [comment]
@MegaManSec rand() a.k.a. @nheir made some nice stuff indeed, the 0.7 BotCTF server is his

**Comment by AssassinTee** [comment]
@MegaManSec The bot of rand() aka. @nheir was client sided while this topic is about server sided bots. Of course you can still use it for this project.

I did some research and there are server sided bots:
Teeworlds 0.6
- zChaos [teeworlds.com entry](https://www.teeworlds.com/forum/viewtopic.php?id=9682), I helped pikotee
- some vanilla bots, link?

Teeworlds 0.7
- Zomb 3 [teeworlds forum entry](https://www.teeworlds.com/forum/viewtopic.php?id=12646) by @LordSk 
- ZoD 2 [teeworlds forum entry](https://www.teeworlds.com/forum/viewtopic.php?id=12724) by myself

**Disclaimer:** The bots in my Mod are **not** nicely written (based on 4 years old code (pre master and bachelor) and I will address this in the future.

Unless I have overseen something, all of this bots can be changed to a [behaviour tree (youtube link)](https://www.youtube.com/watch?v=6VBCXvfNlCM). Behaviour trees are really good, if you want to use a lot of bots. Lots of other AI-algorithms are very CPU intensive (Random Forests, CNNs).

For a good bot support we should split the task in multiple parts (Divide and Conquer). We need:
- A flag in CPlayer, that marks a player as a bot (Dummy?)
- A pathfinding algorithm
- Difficulty flags
    - In case of a behavior tree, you can simply activate/deactive nodes
- An interface to evaluate what the bot does
- An interface for the bot to use the Character
    - simple:
        - Go left/right
        - Shoot at coordinate or player
        - Hook direction/player/coordinate
        - Disconnect
    - advanced:
        - Move Up/Down (combination from hook and left/right)
        - Rocketjump
        - Get flag/Capture flag
        - Heal (move to hearts and/or shields)

I could propose an architecture if you want!

Did I miss something?

Assa

**Comment by nheir** [comment]
you missed the server mod version of my bot (at least)  pointed by dune

[old teeworlds thread](https://www.teeworlds.com/forum/viewtopic.php?id=11594) and [github repo](/nheir/teeworlds)

**Comment by AssassinTee** [comment]
This looks really good, I like how you can draw the path with lasers

**Comment by Dune-jr** [comment]
I think @nheir is a good basis to build on top of. It doesn't have to be state-of-the-art AIs; just simple, dumb bots would achieve a lot. The current level of @nheir's bots is already a reasonable highest difficulty imo

There are plenty of different possible architectures and it's probably best to leave the choice to whoever takes the lead on this task.

**Comment by ChillerDragon** [comment]
>The current level of nheir's bots is already a reasonable highest difficulty imo

Im not sure how much the bot changed since 2016 but back then it was way too weak. I consider myself a weaker player and I managed it back then to beat the bot 10:0 ( https://www.youtube.com/watch?v=4IYfdirJkoY ). So good players would have no challange at all.

But @Speedy-Consoles showed me a clientside bot that was very impressive and even made better players stuggle. But I guess his stuff is closed source and he doesn't do too much tw related stuff anymore. Just wanted to mention that there are good players and it would be sad if they have no bot to train with.

**Comment by heinrich5991** [comment]
I'm sure you can try to contact him for the source.

**Comment by farooqkz** [comment]
What about client bots instead? A client library/module + some AI code.

So anyone interested who knows AI can code bots, from "dummy" bots to "machine learning" bots!
This way it is a lot more flexible and may turn this game into a programming game also! However, bots should have a flag or sign or something which indicates "this client is a bot".

**Comment by jxsl13** [comment]
no, cuz it makes cheating as easy as changing the bot flag from 1 to 0.

**Comment by farooqkz** [comment]
Someone is able to create such a framework and create such a bots anyway...

**Comment by AssassinTee** [comment]
This also comes with an unnecessarry overhead, if the server owner wants to run bots he shouldn't need to run 30 clients for 30 bots

**Comment by jxsl13** [comment]
That's still no argument to actually make it easier for people to create cheats. There are also people who did make AI clients, if they wanted.
